### DISCUSSION TITLE
V8: Replacing the Path type

### DISCUSSION THREAD

**Discussion by felixschorer** [react-hook-form/react-hook-form#7389] [RFC]
# The Problem

`Path` can produce very large union types for nested form data.

## TypeScript Performance
See https://github.com/react-hook-form/react-hook-form/issues/7290.

## Developer Experience
When entering a path, IntelliSense will display the suggested paths in alphabetical order. That results in "deeper" paths being placed above the immediately following paths. 

See this [TS Playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAYg9nAQgQwE4B4AqA+KBeKAbwFgAoKKAIzQC4pMyKAzBOh0gXzLNEljlQBbGAEsIAGwAmAZ3z8kadPAUZlKDNOCoRAOwDm2Q91K9oABWTAAFrIKMoAHygAiaqmf2nrtADo3zqE8XNx8WOACg71Q-X39A8kdg2N8wiISvELCY93iKDJSEUIQPdJdUyKy4isLU3MTnSuSc6rhsovC6ryzG5tKGmoHw40kIAGNxNGgmAFcdUeAROB0oYAhNAGVpvT01haXpAAowSys6C2tpAEo6ADc4EUljVY2tnc1FnUP+8MugA). 

When entering `"foo"` the following suggestions will be shown. 
![image](https://user-images.githubusercontent.com/12783412/147653730-1adc3d4e-d151-434e-a74a-fba9adf831c3.png)
Notice, that `"foo.foo"` is only the 5th suggestion, and that completely unrelated paths are suggested.

# Proposed Solution
## `LazyPath` (Function Arguments)
In places where `Path` is used to describe a function argument, one can use a type which validates the string using [template literal type inference](https://www.typescriptlang.org/docs/handbook/2/template-literal-types.html#inference-with-template-literals).

This has been implemented in this [Code Sandbox](https://codesandbox.io/s/gallant-mountain-ccxz1?file=/src/index.ts).

```diff
- interface SomeComponentProps<T extends FieldValues, P extends Path<T>> {
+ interface SomeComponentProps<T extends FieldValues, P extends string> { 
    control: Control<T>
-   name: P
+   name: LazyPath<T, P>
  }

- function SomeComponent<T extends FieldValues, P extends Path<T>>({ 
+ function SomeComponent<T extends FieldValues, P extends string>({
    control, 
    name,
  }: SomeComponentProps<T, P>) {
    const { field } = useController({ control, name })
    // ...
  }
```

## Other Use Cases
`Path` can be used outside of function arguments. However, `LazyPath` can only be used for function arguments.
For those cases something like https://github.com/react-hook-form/react-hook-form/discussions/7354 could be used.

```diff
  interface TextFieldDescriptor<T extends FieldValues> {
    type: "text"
    label: string
-   path: Path<T>
+   path: PathWithMeta<T, string>
  }

  interface NumberFieldDescriptor<T extends FieldValues> {
    type: "number"
    label: string
-   path: Path<T>
+   path: PathWithMeta<T, number>
  }

  type FormDescriptor<T extends FieldValues> = Array<
    TextFieldDescriptor<T> | NumberFieldDescriptor<T>
  >

  interface FooBar {
    foo: string
    bar: number
  }

  const fooBarForm: FormDescriptor<FooBar> = [
-   { type: 'text', label: 'Foo', path: 'foo' },
-   { type: 'number', label: 'Bar', path: 'bar' },
+   { type: 'text', label: 'Foo', path: of('foo') },
+   { type: 'number', label: 'Bar', path: of('bar') },
  ]
```

cc @barrymay @kotarella1110 @jorisre

**Comment by bluebill1049** [comment]
I would recommend exposing `FieldPath` still (as alias perhaps), as this will be more consistent for exposing type and without confusing with some user's generic type. 

https://react-hook-form.com/ts

**Reply by felixschorer** [reply]
I agree that we should export `FieldPath` as an alias to `Path`(`WithMeta`).

**Reply by bluebill1049** [reply]
> I agree that we should export FieldPath as an alias to Path(WithMeta).

Yea, that's what I am referring to. 👍

**Reply by felixschorer** [reply]
We should consider not exporting `Path` then. Currently `Path` and `FieldPath` is exported.

**Reply by bluebill1049** [reply]
If it's possible, I think it's better to keep those type names relevant to (hook) form.

**Reply by salamaashoush** [reply]
I am currently relying on ` Path`, `PathValue`, Please is it possible to keep those generic types exposed also, and I don't mind also if it is the new lazy variant, I use them to build similar DX as RHF `FieldPath`, Thanks

**Reply by felixschorer** [reply]
@salamaashoush
The current state is that `Path` and `PathValue` are no longer exposed. Instead there's now `FieldPath`, `Auto.FieldPath` and `FieldPathValue` (+ a few variations of these). Despite the name, the generics are no longer constrained to `FieldValues`.
- `FieldPath` requires you to call to `of`, e.g. `of('some.path.to.a.value')`. (formerly `PathWithMeta`)
- `Auto.FieldPath` can only be used on function arguments. (formerly `LazyPath`)
- `FieldPathValue` works with either of these.

If you need the old behavior where `Path` produces an exhaustive list of all paths, then you're currently out of luck.

[We've tried to fix most of the issues with the `Path` type to keep it around, but every time we tried we ran into TS limitations.](https://github.com/react-hook-form/react-hook-form/pull/7368#issuecomment-1001495078)

**Reply by salamaashoush** [reply]
@felixschorer Awesome that's great, I don't need the old behavior I just need the type itself to be exposed without coupling with RHF meta, thanks for clarifying
